proposals/NNNN-unlock-existentials-with-fixed-associated-types.md

```diff
@@ -0,0 +1,253 @@
+
+One may point out, though, that `myUser` could be bound to `some User`, which would have a similar effects. However, [opaque result types](https://github.com/apple/swift-evolution/blob/master/proposals/0244-opaque-result-types.md) are not Existentials. That is, the former allows for hiding type information from the user while the compiler internally retains the underlying type. On the contrary, the latter properly erases type information allowing storage of different types - that conform to a given protocol. What that means, is that should `myUser` be bound to `some User` then initialization would have to be performed at the declaration-site and mutation would be prohibited.
+
+Moreover, we often need to pass around values bound to such protocols through functions and closures. Of course, generics are extremely useful in the case of functions. However, when it comes to closures the user doesn't have many options due to the lack of support for generic arguments:
+
+```swift
+let userProvider: 
+    () -> User
+// ❌ That's invalid.
+    
+
+let userModifier: 
+    <SomeUser: User>(SomeUser) -> SomeUser
+// ❌ Generics are not
+// supported here.
+
+let otherUserProvider: 
+    () -> some User
+// ❌ Neither opaque results 
+// types work here, which  
+// are a form of Generics.
+```
+
+Evidently, this is a great inconvenience with not so elegant workarounds:
+
+```swift
+protocol AnyUser { ... } 
+// Same as `User` above but
+// without `Identifiable` conformance
+
+let myUser: AnyUser
+// ✅
+
+
+protocol User: AnyUser, Identifiable 
+    where ID == String { 
+    
+    var id: String {
+        username
+    }
+}
+
+let myOtherUser: User 
+// ❌ Still an error as 
+// expected
+```
+
+These workarounds, besides producing boilerplate and confusing code, also produce confusing API designs leaving clients to wonder which of the two protocols should be used. In other cases, API authors may decide that inheriting `Identifiable` isn’t worth the added complexity and leave the `User` protocol with no such inheritance whatsoever. 
+
+As a result, a straightforward refinement of simple - yet quite useful - protocols such as `Identifiable` and others becomes a hassle to implement.
+
+
+## Proposed solution
+
+We propose to simply allow `User`-like protocols to be used as Types. Thus, making more natural code possible - which fits a goal of Swift of building ["expressive and elegant APIs"](https://forums.swift.org/t/on-the-road-to-swift-6/32862). Furthermore, library authors could start adding useful protocol conformances to their protocols - a task that's currently prohibitively complex.
+
+
+## Detailed design
+
+### Which Protocols would be Able to be Used as Fully-Fledged Types?
+
+Now, a protocol is usable as a Type when it:
+
+1. Does _not_ include `Self` requirements - and
+2. if it has no associated type requirements _or_ if all its associated types are known.
```

Comment:
I agree with the idea of not using the non-obvious term "Self requirements", however I have a couple of notes on your updated definition:

1. Saying **"the implementations of its associated type requirements are known"** implies that for the rule to apply, the given protocol _must_ have associated requirements; that's not true.

2. (Personal Preference) Saying: **"the implementation of a requirement is known"** may confuse some readers. Maybe, saying: **"its associated types are known"** would suffice. Another direction we could take is use the word "fixed" instead of "known", since the compiler may know the implementation of such a requirement, but still not make it usable as a type due to it not being **fixed**.

3. Saying **"are already preventing it from being supported"** isn't really necessary. It is implied by this definition that whatever doesn't meet the outlined criteria will be unusable as a type.

With all the above changes, the improved definition would look like this:


> Now, a protocol _can_ be used as a type when:
> 1. the implementations of its associated type requirements — _if_ there are any such requirements — are _fixed_, and when
> 2. there are _no_ `Self` references in requirements.

Reply:
> Saying "the implementations of its associated type requirements are known" implies that for the rule to apply, the given protocol must have associated requirements; that's not true.

How about "Now, a protocol with associated type requirements can used as a type when ..."?

> Another direction we could take is use the word "fixed" instead of "known", since the compiler may know the implementation of such a requirement, but still not make it usable as a type due to it not being fixed.

Knowing the implementations of *all* associated type requirements implies that they (the implementations) are fully concrete. "Fixed" is inaccurate for some cases of class-constrained existentials. As soon as I add these examples, the whole set should clarify the possible cases when an associated type requirement can be known on an existential type.

> Saying "are already preventing it from being supported" isn't really necessary. It is implied by this definition that whatever doesn't meet the outlined criteria will be unusable as a type.

We want to avoid explaining "Self references" (not all references to `Self` count). Dropping that part of the sentence makes the criteria opaque.